core/src/main/java/org/apache/iceberg/util/SmallFileUtil.java

```diff
@@ -0,0 +1,145 @@
+import org.slf4j.LoggerFactory;
+
+public class SmallFileUtil {
+  private static final Logger LOG = LoggerFactory.getLogger(SmallFileUtil.class);
+
+  private SmallFileUtil() {
+
+  }
+
+  /**
+   * Returns whether small files should be merged.
+   */
+  public static boolean shouldMergeSmallFiles(Table table) {
+    boolean shouldMerge = false;
+
+    long currentTimeMillis = System.currentTimeMillis();
+    Map<String, String> props = table.properties();
+    boolean autoMergeEnable = PropertyUtil.propertyAsBoolean(
+        props,
+        TableProperties.WRITE_AUTO_COMPACT_ENABLED,
+        TableProperties.WRITE_AUTO_COMPACT_ENABLED_DEFAULT);
+
+    long mergeIntervalMillis = PropertyUtil.propertyAsLong(
+        props,
+        TableProperties.WRITE_COMPACT_INTERVAL_MS,
+        TableProperties.WRITE_COMPACT_INTERVAL_MS_DEFAULT);
+
+    long lastCommittedTimestamp = PropertyUtil.propertyAsLong(
+        props,
+        TableProperties.WRITE_COMPACT_LAST_REWRITE_MS,
+        TableProperties.WRITE_COMPACT_LAST_REWRITE_MS_DEFAULT);
+
+    long smallFileThreshold = PropertyUtil.propertyAsLong(
+        props,
+        TableProperties.WRITE_COMPACT_SMALL_FILE_NUMS,
+        TableProperties.WRITE_COMPACT_SMALL_FILE_NUMS_DEFAULT);
+
+    LOG.info("Summary: actual compact interval: {}s, compact.auto.enabled: {}, " +
+            "compact.interval: {}s, current time: {}, last compact time: {}, small-file-nums: {}",
+        (currentTimeMillis - lastCommittedTimestamp) / 1000,
+        autoMergeEnable,
+        mergeIntervalMillis / 1000,
+        new Date(currentTimeMillis),
+        new Date(lastCommittedTimestamp),
+        smallFileThreshold);
+
+    if (autoMergeEnable && (currentTimeMillis - lastCommittedTimestamp >= mergeIntervalMillis ||
+        smallFileNums(table) >= smallFileThreshold)) {
+      LOG.info("Should compact small files");
+      shouldMerge = true;
+    }
+    return shouldMerge;
+  }
+
+  /**
+   * Returns the number of small files in current snapshot.
+   */
+  public static long smallFileNums(Table table) {
+    Map<String, String> props = table.properties();
+    long smallFileSize = PropertyUtil.propertyAsLong(
+        props,
+        TableProperties.WRITE_COMPACT_SMALL_FILE_SIZE_BYTES,
+        TableProperties.WRITE_COMPACT_SMALL_FILE_SIZE_BYTES_DEFAULT);
+
+    CloseableIterable<FileScanTask> tasks = table.newScan().ignoreResiduals().planFiles();
+    List<DataFile> dataFiles = Lists.newArrayList(CloseableIterable.transform(tasks, FileScanTask::file));
+    List<DataFile> filterDataFiles = dataFiles
+        .stream()
+        .filter(file -> file.fileSizeInBytes() < smallFileSize)
+        .collect(Collectors.toList());
+    LOG.info("small-file-size-bytes: {}, actual small files numbers: {}", smallFileSize, filterDataFiles.size());
+
+    return filterDataFiles.size();
+  }
+
+  /**
+   * Returns whether expire snapshot files should be removed.
+   */
+  public static boolean shouldExpireSnapshot(Table table) {
```

Comment:
In my eyes, this doesn't exactly fit into a "SmallFileUtil". The snapshot expiration is time based, not size based. Does it make sense to move it elsewhere? Possibly there's similar code within the Spark actions for maintaining snapshots that can be leveraged or shared.